### ISSUE TITLE
Is SelfHash agnostic?

### ISSUE THREAD

**Issue by jonasschaefer** [jwkirchenbauer/lm-watermarking#27]
In your readme, you mention that the only downside to the SelfHash scheme is that "this approach requires hashing all possible next tokens". However, I believe that the SelfHash scheme additionally requires any detector to have logit-level model access and thus does not fulfill the _agnostic_ requirement by [(Kuditipudi et al., 2023)](https://arxiv.org/abs/2307.15593). Am I missing something?

**Comment by jwkirchenbauer** [comment]
No, even for SelfHash, we simply run the procedure "as normal", except that the set of token id's being fed into the PRF includes the previous `h` tokens _and_ the token at position `i`. As with the LeftHash version, the outcome of this operation reveals whether the token that was actually sampled at position `i` happened to fall in the green or red set. 

The statement about "hashing all possible next tokens" concerns the task of building the full list of possible next tokens that, if actually selected by the sampler, would satisfy the self-hash condition. In practice, to bound runtime, we resort to just checking the `k` most likely tokens under the model's distribution and building the green set based on which of those satisfy the condition. However, at detection time, we don't need this logits/model distribution over tokens, and need not be aware of whether this iterative process took place.

That said, it's true that the likelihood that the `i`th token sampled happens to be green is _influenced_ by the fact that the only tokens marked as green by the self-hashing procedure, and upweighted by delta as a result, are those that satisfy the condition. Nevertheless, the detection procedure itself doesn't need to recreate the iterative selection process (which would require the model's next token distribution). 

Hope this clarifies things!

**Comment by jonasschaefer** [comment]
Thank you for your excellent answer. I forgot that the detector doesn't recreate the iterative selection process.
